src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -94,25 +94,26 @@ class ContextRDD[C <: AutoCloseable, T: ClassTag](
   def collect(): Array[T] =
     run.collect()
 
+  // WARNING: this resets the context, when this method is called, the value of
+  // type `T` must already be "stable" i.e. not dependent on the region
+  private[this] def decontextualize(c: C, it: Iterator[C => Iterator[T]]): Iterator[T] =
+    new SetupIterator(it.flatMap(_(c)), c.reset _)
+
   def run[U >: T : ClassTag]: RDD[U] =
-    rdd.mapPartitions { part => using(mkc()) { cc => part.flatMap(_(cc)) } }
+    rdd.mapPartitions { part => using(mkc()) { cc => decontextualize(cc, part) } }
 
   private[this] def inCtx[U: ClassTag](
     f: C => Iterator[U]
   ): Iterator[C => Iterator[U]] = Iterator.single(f)
 
-  private[this] def withoutContext[U: ClassTag](
-    f: Iterator[T] => Iterator[U]
-  ): ContextRDD[C, U] = new ContextRDD(rdd.map(_.andThen(f)), mkc)
-
   def map[U: ClassTag](f: T => U): ContextRDD[C, U] =
-    withoutContext(_.map(f))
+    mapPartitions(_.map(f), true)
```

Comment:
IntelliJ and I prefer including the parameter name on Boolean literals: `preservesPartitioning = true`. Makes it easier to understand the code.